source/common/formatter/substitution_formatter.cc

```diff
@@ -54,5 +54,97 @@ const std::regex& SubstitutionFormatParser::commandWithArgsRegex() {
   // clang-format on
 }
 
+JsonFormatBuilder::FormatterEelements
+JsonFormatBuilder::fromStruct(const ProtobufWkt::Struct& struct_format) const {
+  sanitize_buffer_.clear();
+  raw_json_string_.clear();
+  output_.clear();
+
+  formatValueToFormatterEelements(struct_format);
+  if (!raw_json_string_.empty()) {
+    output_.push_back(JsonString{std::move(raw_json_string_)});
+  }
+  return std::move(output_);
+};
+
+void JsonFormatBuilder::formatValueToFormatterEelements(const ProtobufWkt::Value& value) const {
+  switch (value.kind_case()) {
+  case ProtobufWkt::Value::KIND_NOT_SET:
+  case ProtobufWkt::Value::kNullValue:
+    raw_json_string_.append(NullValue);
+    break;
+  case ProtobufWkt::Value::kNumberValue:
+    if (keep_value_type_) {
+      // Keep the number type in JSON.
+      raw_json_string_.append(fmt::to_string(value.number_value()));
+    } else {
+      // Convert the number to string.
+      raw_json_string_.append(fmt::format(R"("{}")", value.number_value()));
+    }
+    break;
+  case ProtobufWkt::Value::kStringValue: {
+    absl::string_view string_format = value.string_value();
+    if (string_format.empty() || !absl::StrContains(string_format, '%')) {
+      raw_json_string_.append(
+          fmt::format(R"("{}")", Json::sanitize(sanitize_buffer_, string_format)));
+    } else {
+      // The string contains a formatter, we need to push the current raw string
+      // into the output list first.
+      if (!raw_json_string_.empty()) {
+        output_.push_back(JsonString{std::move(raw_json_string_)});
+      }
+
+      // Now a formatter is coming, we need to push the current raw string into
+      // the output list.
+      output_.push_back(TmplString{std::string(string_format)});
+    }
+    break;
+  }
+  case ProtobufWkt::Value::kBoolValue:
+    if (keep_value_type_) {
+      // Keep the bool type in JSON.
+      raw_json_string_.append(fmt::to_string(value.bool_value()));
+    } else {
+      // Convert the bool to string.
+      raw_json_string_.append(fmt::format(R"("{}")", value.bool_value()));
```

Comment:
for these cases where the only difference is whether the value is surrounded by quotes, can we use this pattern?

appendValue(computeFormattedString(...));

with 
```
void JsonFormatBuilder::appendValue(absl::string_view value) {
  if (keep_value_type_) {
    raw_json_str_.append(value);
  } else {
    absl::StrAppend(&raw_json_str, "\"", value, "\"");
  }
}
```

I think that would (a) centrallize the messy code to add optional quotes (b) be faster because we wouldn't be having fmt::format parse a format string at runtime for every element.

Reply:
Great suggestion.

There code is used to load the configuration and won't effect the core data paths, so I didn't optimize it heavily in performance. But it's great to centralize the messy code.